pkg/xds/server/callbacks/dataplane_sync_callbacks.go

```diff
@@ -0,0 +1,230 @@
+	proxyInfo *dpInfo
+}
+
+func (t *dataplaneSyncCallbacks) OnStreamResponse(i int64, request util_xds.DiscoveryRequest, response util_xds.DiscoveryResponse) {
+	// Noop
+}
+
+func (t *dataplaneSyncCallbacks) OnStreamOpen(ctx context.Context, streamID core_xds.StreamID, _ string) error {
+	t.Lock()
+	defer t.Unlock()
+	dataplaneSyncLog.V(1).Info("stream is open", "streamID", streamID)
+	if _, found := t.activeStreams[streamID]; found {
+		return errors.Errorf("streamID %d is already tracked, we should never reopen a stream", streamID)
+	}
+	t.activeStreams[streamID] = &streamInfo{ctx: ctx, proxyInfo: nil}
+	return nil
+}
+
+func (t *dataplaneSyncCallbacks) OnStreamRequest(streamID core_xds.StreamID, request util_xds.DiscoveryRequest) error {
+	if request.NodeId() == "" {
+		// from https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#ack-nack-and-versioning:
+		// Only the first request on a stream is guaranteed to carry the node identifier.
+		// The subsequent discovery requests on the same stream may carry an empty node identifier.
+		// This holds true regardless of the acceptance of the discovery responses on the same stream.
+		// The node identifier should always be identical if present more than once on the stream.
+		// It is sufficient to only check the first message for the node identifier as a result.
+		return nil
+	}
+
+	t.RLock()
+	dataplaneSyncLog.V(1).Info("stream request", "streamID", streamID)
+	alreadyProcessed := t.activeStreams[streamID].proxyInfo != nil
+	t.RUnlock()
+	if alreadyProcessed { // We fast return if we already know that this streamID is tracking a specific dataplane
+		return nil
+	}
+
+	proxyId, err := core_xds.ParseProxyIdFromString(request.NodeId())
+	if err != nil {
+		return errors.Wrap(err, "invalid node ID")
+	}
+	dpKey := proxyId.ToResourceKey()
+	metadata := core_xds.DataplaneMetadataFromXdsMetadata(request.Metadata())
+	if metadata == nil {
+		return errors.New("metadata in xDS Node cannot be nil")
+	}
+	if metadata.Resource != nil {
+		if core_model.MetaToResourceKey(metadata.Resource.GetMeta()) != dpKey {
+			return fmt.Errorf("proxyId %s does not match proxy resource %s", dpKey, metadata.Resource.GetMeta())
+		}
+	}
+	l := dataplaneSyncLog.WithValues("dpKey", dpKey, "streamID", streamID)
+	t.Lock()
+	if t.activeStreams[streamID].proxyInfo != nil {
+		return nil // We fast return if we already know that this streamID is tracking a specific dataplane
+	}
+	pInfo := t.proxyInfos[dpKey]
+	if pInfo == nil {
+		ctx, cancel := context.WithCancel(context.Background())
+		done := make(chan struct{})
+		pInfo = &dpInfo{
+			dpKey:            dpKey,
+			lifecycleManager: t.dpLifecycleManagerFactory.New(dpKey),
+			ctx:              ctx,
+			done:             done,
+			meta:             atomic.Pointer[core_xds.DataplaneMetadata]{},
+			cancelFunc:       cancel,
+		}
+		t.proxyInfos[dpKey] = pInfo
+	}
+	t.activeStreams[streamID].proxyInfo = pInfo
+	ctx := t.activeStreams[streamID].ctx
+	pInfo.Lock()
```

Comment:
Actually the fact that the long running cleanup runs with this lock means that 2 connections to the same DP may lead to holding the global lock for longer.